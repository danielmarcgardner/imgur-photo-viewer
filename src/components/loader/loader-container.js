import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import Loader from './loader';
import { searchForImages } from '../../redux/actions/thunks';

export const mapStateToProps = state => ({
  fetching: state.searchResults.fetching,
  searchTerm: state.searchResults.searchTerm,
  noImagesFound: state.searchResults.noImages,
  error: state.searchResults.error,
});

export const mapDispatchToProps = dispatch => bindActionCreators({
  searchForImages,
}, dispatch);

export default connect(mapStateToProps, mapDispatchToProps)(Loader);
