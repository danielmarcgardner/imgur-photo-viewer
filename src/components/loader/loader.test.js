import React from 'react';
import { shallow, configure, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import Loader from './loader';
import { mapStateToProps, mapDispatchToProps } from './loader-container';

configure({ adapter: new Adapter() });

const searchForImages = jest.fn();

describe('Loader', () => {
  it('Renders to the page when not fetching', () => {
    const component = shallow(
      <Loader
        fetching={false}
        searchForImages={searchForImages}
        searchTerm={'bulldog'}
        noImagesFound={false}
      />,
    );
    expect(component).toMatchSnapshot();
  });
  it('Renders to the page when fetching', () => {
    const component = shallow(
      <Loader
        fetching
        searchForImages={searchForImages}
        searchTerm={'bulldog'}
        noImagesFound={false}
      />,
    );
    expect(component).toMatchSnapshot();
  });
  it('Renders to page when no results', () => {
    const component = shallow(
      <Loader
        fetching={false}
        searchForImages={searchForImages}
        searchTerm={'asdfghjkjhgfdsdfgh'}
        noImagesFound
      />,
    );
    expect(component).toMatchSnapshot();
  });
  it('Renders with correct properties', () => {
    const componentFetching = shallow(
      <Loader
        fetching
        searchForImages={searchForImages}
        searchTerm={'bulldog'}
        noImagesFound={false}
      />,
    );
    const component = shallow(
      <Loader
        fetching={false}
        searchForImages={searchForImages}
        searchTerm={'bulldog'}
        noImagesFound={false}
      />,
    );
    const componentNoImages = shallow(
      <Loader
        fetching={false}
        searchForImages={searchForImages}
        searchTerm={'asdfghjkjhgfdsdfgh'}
        noImagesFound
      />,
    );
    expect(componentFetching.find('#Loading').length).toBe(1);
    expect(component.find('button').length).toBe(1);
    expect(componentNoImages.find('#NotFound').length).toBe(1);
  });
  it('When clicked searchForImages is called with the correct search term', () => {
    const component = mount(
      <Loader
        fetching={false}
        searchForImages={searchForImages}
        searchTerm={'bulldog'}
        noImagesFound={false}
      />,
    );
    component.find('button').simulate('click');
    expect(searchForImages).toHaveBeenCalled();
    expect(searchForImages.mock.calls[0][0]).toBe('bulldog');
  });
  it('mapStateToProps returns the correct state', () => {
    const state = {
      searchResults: {
        images: [
          {
            image: 'https://i.imgur.com/fZ8JNw4.jpg',
            title: 'Skateboarding Bulldog',
            tags: ['Funny', 'Dog', 'Action'],
            score: 100,
            datetime: 'Mon Oct 02 2017',
            id: 'fZ8JNw4',
          },
          {
            id: 'gzC6hl4',
            image: 'https://i.imgur.com/gzC6hl4.gif',
            title: 'Tired Bulldog Dad Needs Break From His Playful Puppies.',
            score: 39,
            tags: [],
            datetime: 'Mon Oct 02 2017',
          }, {
            id: 'xE7Yu2v',
            image: 'https://i.imgur.com/xE7Yu2v.jpg',
            title: 'Trevor Donovan with his bulldog Tito and German Shepherd Dogbert.',
            score: 26,
            tags: ['dogs', 'dog', 'german sheâ€¦', 'pets'],
            datetime: 'Mon Oct 02 2017' },
        ],
        searchTerm: 'bulldog',
        fetching: false,
        noImages: false,
      },
    };
    const expected = {
      searchTerm: 'bulldog',
      fetching: false,
      noImagesFound: false,
    };
    const actual = mapStateToProps(state);
    expect(actual).toEqual(expected);
  });
  it('mapDispatchToProps dispatches the expected action', () => {
    const dispatch = jest.fn();
    const dispatched = mapDispatchToProps(dispatch);
    expect(dispatched).toHaveProperty('searchForImages');
  });
});
