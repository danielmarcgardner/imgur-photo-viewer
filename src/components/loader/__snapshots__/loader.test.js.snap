// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Loader Renders to page when an error occurs 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Loader
    error={true}
    fetching={false}
    noImagesFound={false}
    searchForImages={[Function]}
    searchTerm="asdfghjkjhgfdsdfgh"
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": null,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <img
          alt="Error Occured"
          src="warning-icon.svg"
/>,
        <p>
          Sorry an unexpected error has occured. Please try a new search!
</p>,
      ],
      "className": "NotFound",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": null,
        "nodeType": "host",
        "props": Object {
          "alt": "Error Occured",
          "src": "warning-icon.svg",
        },
        "ref": null,
        "rendered": null,
        "type": "img",
      },
      Object {
        "instance": null,
        "key": null,
        "nodeType": "host",
        "props": Object {
          "children": "Sorry an unexpected error has occured. Please try a new search!",
        },
        "ref": null,
        "rendered": "Sorry an unexpected error has occured. Please try a new search!",
        "type": "p",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": null,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <img
            alt="Error Occured"
            src="warning-icon.svg"
/>,
          <p>
            Sorry an unexpected error has occured. Please try a new search!
</p>,
        ],
        "className": "NotFound",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": null,
          "nodeType": "host",
          "props": Object {
            "alt": "Error Occured",
            "src": "warning-icon.svg",
          },
          "ref": null,
          "rendered": null,
          "type": "img",
        },
        Object {
          "instance": null,
          "key": null,
          "nodeType": "host",
          "props": Object {
            "children": "Sorry an unexpected error has occured. Please try a new search!",
          },
          "ref": null,
          "rendered": "Sorry an unexpected error has occured. Please try a new search!",
          "type": "p",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`Loader Renders to page when no results 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Loader
    error={false}
    fetching={false}
    noImagesFound={true}
    searchForImages={[Function]}
    searchTerm="asdfghjkjhgfdsdfgh"
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": null,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <img
          alt="Cannot find search results"
          src="warning-icon.svg"
/>,
        <p>
          Sorry we can't find any more images with that search. Please try a new search!
</p>,
      ],
      "className": "NotFound",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": null,
        "nodeType": "host",
        "props": Object {
          "alt": "Cannot find search results",
          "src": "warning-icon.svg",
        },
        "ref": null,
        "rendered": null,
        "type": "img",
      },
      Object {
        "instance": null,
        "key": null,
        "nodeType": "host",
        "props": Object {
          "children": "Sorry we can't find any more images with that search. Please try a new search!",
        },
        "ref": null,
        "rendered": "Sorry we can't find any more images with that search. Please try a new search!",
        "type": "p",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": null,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <img
            alt="Cannot find search results"
            src="warning-icon.svg"
/>,
          <p>
            Sorry we can't find any more images with that search. Please try a new search!
</p>,
        ],
        "className": "NotFound",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": null,
          "nodeType": "host",
          "props": Object {
            "alt": "Cannot find search results",
            "src": "warning-icon.svg",
          },
          "ref": null,
          "rendered": null,
          "type": "img",
        },
        Object {
          "instance": null,
          "key": null,
          "nodeType": "host",
          "props": Object {
            "children": "Sorry we can't find any more images with that search. Please try a new search!",
          },
          "ref": null,
          "rendered": "Sorry we can't find any more images with that search. Please try a new search!",
          "type": "p",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`Loader Renders to the page when fetching 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Loader
    error={false}
    fetching={true}
    noImagesFound={false}
    searchForImages={[Function]}
    searchTerm="bulldog"
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": null,
    "nodeType": "host",
    "props": Object {
      "children": "Loading...",
      "id": "Loading",
    },
    "ref": null,
    "rendered": "Loading...",
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": null,
      "nodeType": "host",
      "props": Object {
        "children": "Loading...",
        "id": "Loading",
      },
      "ref": null,
      "rendered": "Loading...",
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`Loader Renders to the page when not fetching 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Loader
    error={false}
    fetching={false}
    noImagesFound={false}
    searchForImages={[Function]}
    searchTerm="bulldog"
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": null,
    "nodeType": "host",
    "props": Object {
      "children": <button
        onClick={[Function]}
>
        Load More!
</button>,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": null,
      "nodeType": "host",
      "props": Object {
        "children": "Load More!",
        "onClick": [Function],
      },
      "ref": null,
      "rendered": "Load More!",
      "type": "button",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": null,
      "nodeType": "host",
      "props": Object {
        "children": <button
          onClick={[Function]}
>
          Load More!
</button>,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": null,
        "nodeType": "host",
        "props": Object {
          "children": "Load More!",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": "Load More!",
        "type": "button",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
